File: generated-code/codecraft/rust/src/model/entity_type.rs
use super::*;

/// Entity type
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub enum EntityType {
    /// Wall, can be used to prevent enemy from moving through
    Wall,
    /// House, used to increase population
    House,
    /// Base for recruiting new builder units
    BuilderBase,
    /// Builder unit can build buildings
    BuilderUnit,
    /// Base for recruiting new melee units
    MeleeBase,
    /// Melee unit
    MeleeUnit,
    /// Base for recruiting new ranged units
    RangedBase,
    /// Ranged unit
    RangedUnit,
    /// Resource can be harvested
    Resource,
    /// Ranged attacking building
    Turret,
}

impl trans::Trans for EntityType {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        let tag: i32 = match self {
            Self::Wall => 0,
            Self::House => 1,
            Self::BuilderBase => 2,
            Self::BuilderUnit => 3,
            Self::MeleeBase => 4,
            Self::MeleeUnit => 5,
            Self::RangedBase => 6,
            Self::RangedUnit => 7,
            Self::Resource => 8,
            Self::Turret => 9,
        };
        trans::Trans::write_to(&tag, writer)
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let tag = <i32 as trans::Trans>::read_from(reader)?;
        match tag {
            0 => Ok(Self::Wall),
            1 => Ok(Self::House),
            2 => Ok(Self::BuilderBase),
            3 => Ok(Self::BuilderUnit),
            4 => Ok(Self::MeleeBase),
            5 => Ok(Self::MeleeUnit),
            6 => Ok(Self::RangedBase),
            7 => Ok(Self::RangedUnit),
            8 => Ok(Self::Resource),
            9 => Ok(Self::Turret),
            _ => Err(std::io::Error::new(
                std::io::ErrorKind::Other,
                format!("Unexpected tag {:?}", tag))),
        }
    }
}

File: generated-code/codecraft/swift/TcpReadWrite/Sources/TransGenTest/EntityType.swift
/// Entity type
public enum EntityType: Int32 {
    /// Wall, can be used to prevent enemy from moving through
    case wall = 0

    /// House, used to increase population
    case house = 1

    /// Base for recruiting new builder units
    case builderBase = 2

    /// Builder unit can build buildings
    case builderUnit = 3

    /// Base for recruiting new melee units
    case meleeBase = 4

    /// Melee unit
    case meleeUnit = 5

    /// Base for recruiting new ranged units
    case rangedBase = 6

    /// Ranged unit
    case rangedUnit = 7

    /// Resource can be harvested
    case resource = 8

    /// Ranged attacking building
    case turret = 9

    /// Read EntityType from input stream
    static func readFrom<S: InputStream>(_ stream: S) -> EntityType {
        return EntityType(rawValue: stream.readInt32())!
    }

    /// Write EntityType to output stream
    func writeTo<S: OutputStream>(_ stream: S) {
        stream.writeInt32(rawValue)
    }
}